### PR TITLE
Feat: PUT and GET organization and tests

### LINKED ISSUE #105: Dev: Get and Create/Update organization profile
## Description
As an organization user,
I want to create/update organization profile,
so that I can modify details about organization.

## Acceptance Criteria
- [ ] GET request to get the organization profile
- [ ] PUT request to update the organization profile 
- [ ] Handle bad request if any payload is missing, irrelevant
- [ ] Handle authorization

## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 2 mentor.

### PR DESCRIPTION
**Pull request by mtreacy002**
### Description
Create PUT and GET /organization API endpoints

Fixes #105 

### Type of Change:
- Code

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
* Run and test the application manually:
- Login as user who already has additional_info in the database and `is_organization_rep` is True
- try to send PUT /organization request before setting Cookie `user_id` through GET /organization. see error 403 Forbidden and error message
Screen Shot 2020-08-11 at 9 02 35 pm

- send GET /organization request. Since this user has not created an organization, see error 404 Not Found and message `Organization does not exist`.
Screen Shot 2020-08-11 at 1 08 11 am

- now create organization by sending PUT /organization request with correct payload. see code 201 Created and success message
Screen Shot 2020-08-11 at 9 30 29 pm

- now try to update the organization by changing some details. On correct update, see code 200 and success message
Screen Shot 2020-08-11 at 9 34 01 pm

- logout and login as a user that either has additional_info `is_organization_rep` is `False` or has not created an additional_info. 
- send GET /organization request. see error 403 Forbidden and error message user has not declared `is_organization_rep` value
Screen Shot 2020-08-11 at 1 16 48 am

- an example of successful GET /organization request will be as below
Screen Shot 2020-08-12 at 11 38 28 pm

* Run unittests and ensure all tests passed

### Checklist:
- [ ] My PR follows the style guidelines of this project
- [ ] I have performed a self-review of my own code or materials
- [ ] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] Any dependent changes have been merged

**Code/Quality Assurance Only**
- [ ] My changes generate no new warnings
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

**Additional Info**
This is still a work in progress

### CODE REVIEW

**Comment by mtreacy002** [comment]
Update @anitab-org/bridgeintech-maintainers . This PR is still a WIP. At the moment it addresses GET /organization with responses 403 Forbidden when the user either did not declare that they represent an organization or they haven't created an additional information instance to state whether or not they represents an organisation.

**Comment by mtreacy002** [comment]
@ramitsawhney27 . Can you please check this PR and give me feedback whether or not my approach on this GET /organization make sense. Thanks

**Comment by mtreacy002** [comment]
@ramitsawhney27, I specifically want to note that I send an HTTP request to GET /user to backend within the logic of GET /organization since to get the organization that the logged-in user represents, I need to pass the user_id as parameter in the database query.

**Comment by mtreacy002** [comment]
Update @anitab-org/bridgeintech-maintainers . I've completed the GET and PUT API endpoints of this PR. I still need to do the test cases though. Will force push it once I'm done. @ramitsawhney27 , can you please open an issue for this functionalities (GET and PUT /organizations) so I can link this PR to it. Thanks

**Comment by ramitsawhney27** [comment]
> @ramitsawhney27 . Can you please check this PR and give me feedback whether or not my approach on this GET /organization make sense. Thanks

This is fine

**Comment by mtreacy002** [comment]
Update @anitab-org/bridgeintech-maintainers . This PR is now completed. Please review when you have time. Thanks

File: app/api/dao/organization.py
Comment (meenakshi-dhanani) [inline app/api/dao/organization.py]: Maya it's a good thing to not use abbreviations. It gets really difficult when you're code is worked on by more people. Try not using abbreviations anywhere, could be `representative_id`.

File: app/api/dao/organization.py
Comment (meenakshi-dhanani) [inline app/api/dao/organization.py]: what does it mean? IS_ORGANIZATION_REP_NOT_DECLARED? can we name this better?

Comment (meenakshi-dhanani) [reply app/api/dao/organization.py]: https://www.codingblocks.net/podcast/clean-code-writing-meaningful-names/ - whenever you're free, Uncle Bob's clean code posts/podcasts are available. They should explain how important it is to have meaningful names

File: app/api/dao/organization.py
Comment (mtreacy002) [reply app/api/dao/organization.py]: ok, will change it.

File: app/api/dao/organization.py
Comment (mtreacy002) [reply app/api/dao/organization.py]: > what does it mean? IS_ORGANIZATION_REP_NOT_DECLARED? can we name this better?

This means the user either has not declared whether or not they are an organization representative or has not yet created an additional_info and declare they are a representative. Tbh I can't think of any other name that is more suitable to be the title for this message. Do you have a suggestion @meenakshi-dhanani ?

File: app/api/dao/organization.py
Comment (ramitsawhney27) [inline app/api/dao/organization.py:29]: Can this be EAFP/try-catch based? Makes more sense to me intuitively

File: app/api/dao/organization.py
Comment (ramitsawhney27) [inline app/api/dao/organization.py:71]: +1 for try-except

File: app/api/dao/organization.py
Comment (mtreacy002) [reply app/api/dao/organization.py:29]: ok, will try and change it now.

Comment (mtreacy002) [reply app/api/dao/organization.py:29]: @ramitsawhney27 , I'm not sure how to use EAFP to check if the value of is_organization_rep is true, which is my intention here. Can you please advise?

Comment (ramitsawhney27) [reply app/api/dao/organization.py:29]: I meant for line 30-44. We can do:

try: 
    return {}
except ...:
    return messages. OR...

Not a compulsory thing, if you're unclear or disagree, we can skip this for now

Comment (mtreacy002) [reply app/api/dao/organization.py:29]: I see. Done. I've changed it to EAFP now 😁

**Review by ramitsawhney27 — DISMISSED** [review]

**Comment by mtreacy002** [comment]
Update @ramitsawhney27 and @anitab-org/bridgeintech-maintainers. I've pushed a minor change to GET /organization so that it returns result if first try is successful. This way it's less code as well (remove unnecessary check on user_id).

**Comment by mtreacy002** [comment]
Update @anitab-org/bridgeintech-maintainers . I need to refactor Organization dao get response model so that it returns representative name as well as ID to make it meaningful, in particular later since this will be used for viewing Organizations list on UI. So, please do not merge this PR for now.

**Comment by mtreacy002** [comment]
Update @ramitsawhney27 and @anitab-org/bridgeintech-maintainers .I've pushed the latest change which refactored OrganizationDAO get organization response model to include representative name. I've also modified the tests accordingly. Please re-review when you have time.

**Review by ramitsawhney27 — DISMISSED** [review]

File: app/api/dao/organization.py
Comment (meenakshi-dhanani) [inline app/api/dao/organization.py]: this rep_id -> representative_id?

File: app/api/dao/organization.py
Comment (meenakshi-dhanani) [reply app/api/dao/organization.py]: NO_REPRESENTATIVE_FOR_ORGANIZATION?

File: app/messages.py
Comment (meenakshi-dhanani) [inline app/messages.py]: I think remove the You can do this. It's really upto change, if we change any structure on the UI, we'll always have to remember to change this text. 
Too tightly coupled

File: tests/organizations/test_api_get_organization.py
Comment (meenakshi-dhanani) [inline tests/organizations/test_api_get_organization.py:83]: don't use rep please. Remove all reps, make them representative or whatever the full form is. Somebody besides you is going to have a real hard time figuring what rep is

**Comment by meenakshi-dhanani** [comment]
For a user, that is registered and has not declared their organization I posted this in the payload: 
```
{
  "rep_department": "HR",
  "name": "ThoughtWorks",
  "email": "tw@123.com",
  "about": "Great org",
  "address": "Somewhere",
  "website": "tw.com",
  "timezone": "",
  "phone": "8989080800",
  "status": "single"
}
```

and I got this response with the following message: 

```
403:   "message": "You must send GET /organization request first before you can proceed."

```

I thought 403 would say I haven't declared organization or something

File: app/api/dao/organization.py
Comment (mtreacy002) [reply app/api/dao/organization.py]: Ok, will change it.

File: tests/organizations/test_api_get_organization.py
Comment (mtreacy002) [reply tests/organizations/test_api_get_organization.py:83]: this one is unavoidable (as per the other remaining rep_.. that is part of database attribute). I have changed the models and other that not related to the database. If you still insist on the rep_ to be changed on the database model, I need to redo not only the Organization and Program models, but also recreate tables on the database. The samee need to be done with the remote database on Heroku. 
So, do we wanna go that far with the remaining of GSoC (restructuring the database)? Or can we put this in icebox to be pick up after GSoC. 
PS: Before answering this, keep in mind I still have lots of deliverables untouched such as the Program, Applying to programs, Mentorship relation, static pages for Homepage (About, Benefit, Contact), Portfolios (User, Organization and Program) and My Spcace Request History.

**Comment by mtreacy002** [comment]
> For a user, that is registered and has not declared their organization I posted this in the payload:
> 
> ```
> {
>   "rep_department": "HR",
>   "name": "ThoughtWorks",
>   "email": "tw@123.com",
>   "about": "Great org",
>   "address": "Somewhere",
>   "website": "tw.com",
>   "timezone": "",
>   "phone": "8989080800",
>   "status": "single"
> }
> ```
> 
> and I got this response with the following message:
> 
> ```
> 403:   "message": "You must send GET /organization request first before you can proceed."
> ```
> 
> I thought 403 would say I haven't declared organization or something

@meenakshi-dhanani . 403 Forbidden for PUT also return when the user has not send GET /organization as per mentioned in the API endpoint description
Screen Shot 2020-08-13 at 4 31 19 pm

Also can be seen in the code
Screen Shot 2020-08-13 at 4 30 57 pm

This is because through GET /organization, the cookiee for `user_id` is set. This is mainly to inform backend developer since on the frontend UI, when user navigate to the Organization profile page, the GET /organization will be sent first before the user can continue with Update or Create (PUT /organization). Hope this makes sense.

**Comment by meenakshi-dhanani** [comment]
> > For a user, that is registered and has not declared their organization I posted this in the payload:
> > ```
> > {
> >   "rep_department": "HR",
> >   "name": "ThoughtWorks",
> >   "email": "tw@123.com",
> >   "about": "Great org",
> >   "address": "Somewhere",
> >   "website": "tw.com",
> >   "timezone": "",
> >   "phone": "8989080800",
> >   "status": "single"
> > }
> > ```
> > 
> > 
> > and I got this response with the following message:
> > ```
> > 403:   "message": "You must send GET /organization request first before you can proceed."
> > ```
> > 
> > 
> > I thought 403 would say I haven't declared organization or something
> 
> @meenakshi-dhanani . 403 Forbidden for PUT also return when the user has not send GET /organization as per mentioned in the API endpoint description
> Screen Shot 2020-08-13 at 4 31 19 pm
> 
> Also can be seen in the code
> Screen Shot 2020-08-13 at 4 30 57 pm
> 
> This is because through GET /organization, the cookiee for `user_id` is set. This is mainly to inform backend developer since on the frontend UI, when user navigate to the Organization profile page, the GET /organization will be sent first before the user can continue with Update or Create (PUT /organization). Hope this makes sense.

So I cannot test this API on this branch? I'll have to get the branch for the get request and then use it?

**Comment by meenakshi-dhanani** [comment]
There's also another problem, I updated the additional info to make representative true. And it breaks with 
```
TypeError: Failed to fetch
```

I guess because I didn't do the GET? Either way this should be handled gracefully ?

**Comment by mtreacy002** [comment]
> There's also another problem, I updated the additional info to make representative true. And it breaks with
> 
> ```
> TypeError: Failed to fetch
> ```
> 
> I guess because I didn't do the GET? Either way this should be handled gracefully ?

GET and PUT are on the same branch, so you can test it on this branch

**Comment by meenakshi-dhanani** [comment]
> > There's also another problem, I updated the additional info to make representative true. And it breaks with
> > ```
> > TypeError: Failed to fetch
> > ```
> > 
> > 
> > I guess because I didn't do the GET? Either way this should be handled gracefully ?
> 
> GET and PUT are on the same branch, so you can test it on this branch

Can you check on the PUT call, for a user who has represented marked as true, without making the GET call if he/she does the PUT call, what gets returned? For me it gives type error, if it's the same for you can you handle it gracefully maybe 403 the same way you're doing for the other case?

**Comment by mtreacy002** [comment]
> There's also another problem, I updated the additional info to make representative true. And it breaks with
> 
> ```
> TypeError: Failed to fetch
> ```
> 
> I guess because I didn't do the GET? Either way this should be handled gracefully ?

If we look at the fact that our user is only interacting through Frontend UI, instead of Swagger UI (as I mentioned [above](https://github.com/anitab-org/bridge-in-tech-backend/pull/103#issuecomment-673291501)), we don't have to worry that this issue will ever come up (that user forgot to send GET /organization as backend developer). Because as I mentioned [above](https://github.com/anitab-org/bridge-in-tech-backend/pull/103#issuecomment-673291501), when the user click Organization profile on UI, before the page is rendered, we send GET /organization to backend. 
Again, the method I put in place here is to prevent backend developer from not sending PUT before GET.

I put the GET /organization as a blocker for 2 reasons:
1. We need to set the auth cookie user_id which only can be set by sending MS GET /user request. 
2. Since to get to Organization page UI, we'll use GET /organization anyway, so I thought might as well use this endpoint to also send MS GET /user request.

Another place where I have called MS GET /user to set cookie user_id is on GET /user/personal_details for My Space pages. As this might be used often, I wonder if it is best to keep the user instance as part of cookie at the beginning when user is login (so we send POST /login request as well as GET /user), this way we can minimize the need for sending it on both GET /user/personal_details and GET /organization.

**Comment by meenakshi-dhanani** [comment]
> > There's also another problem, I updated the additional info to make representative true. And it breaks with
> > ```
> > TypeError: Failed to fetch
> > ```
> > 
> > 
> > I guess because I didn't do the GET? Either way this should be handled gracefully ?
> 
> If we look at the fact that our user is only interacting through Frontend UI, instead of Swagger UI (as I mentioned [above](https://github.com/anitab-org/bridge-in-tech-backend/pull/103#issuecomment-673291501)), we don't have to worry that this issue will ever come up (that user forgot to send GET /organization as backend developer). Because as I mentioned [above](https://github.com/anitab-org/bridge-in-tech-backend/pull/103#issuecomment-673291501), when the user click Organization profile on UI, before the page is rendered, we send GET /organization to backend.
> Again, the method I put in place here is to prevent backend developer from not sending PUT before GET.
> 
> I put the GET /organization as a blocker for 2 reasons:
> 
> 1. We need to set the auth cookie user_id which only can be set by sending MS GET /user request.
> 2. Since to get to Organization page UI, we'll use GET /organization anyway, so I thought might as well use this endpoint to also send MS GET /user request.
> 
> Another place where I have called MS GET /user to set cookie user_id is on GET /user/personal_details for My Space pages. As this might be used often, I wonder if it is best to keep the user instance as part of cookie at the beginning when user is login (so we send POST /login request as well as GET /user), this way we can minimize the need for sending it on both GET /user/personal_details and GET /organization.

Sounds like a better option

**Comment by meenakshi-dhanani** [comment]
Another thing, so I'm doing the GET now and it says-Internal Server Error when I hit get,
```
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
  "http://www.w3.org/TR/html4/loose.dtd">
<html>
  <head>
    <title>NameError: name 'logging' is not defined // Werkzeug Debugger</title>
    <link rel="stylesheet" href="?__debugger__=yes&amp;cmd=resource&amp;f=style.css"
        type="text/css">
    <!-- We need to make sure this has a favicon so that the debugger does
         not by accident trigger a request to /favicon.ico which might
         change the application state. -->
    <link rel="shortcut icon"
        href="?__debugger__=yes&amp;cmd=resource&amp;f=console.png">
    <script src="?__debugger__=yes&amp;cmd=resource&amp;f=jquery.js"></script>
    <script src="?__debugger__=yes&amp;cmd=resource&amp;f=debugger.js"></script>
    <script type="text/javascript">
      var TRACEBACK = 4500959440,
          CONSOLE_MODE = false,
          EVALEX = true,
          EVALEX_TRUSTED = false,
          SECRET = "hOCDEGO6cGIknxA0j847";
    </script>
  </head>
  <body style="background-color: #fff">
    <div class="debugger">
<h1>NameError</h1>
<div class="detail">
```

**Comment by mtreacy002** [comment]
> > > There's also another problem, I updated the additional info to make representative true. And it breaks with
> > > ```
> > > TypeError: Failed to fetch
> > > ```
> > > 
> > > 
> > > I guess because I didn't do the GET? Either way this should be handled gracefully ?
> > 
> > 
> > GET and PUT are on the same branch, so you can test it on this branch
> 
> Can you check on the PUT call, for a user who has represented marked as true, without making the GET call if he/she does the PUT call, what gets returned? For me it gives type error, if it's the same for you can you handle it gracefully maybe 403 the same way you're doing for the other case?

@meenakshi-dhanani . I don't get the same error as you did. I've got the expected 403 GET /organization not sent.

Screen Shot 2020-08-13 at 5 16 01 pm

**Comment by meenakshi-dhanani** [comment]
> > > > There's also another problem, I updated the additional info to make representative true. And it breaks with
> > > > ```
> > > > TypeError: Failed to fetch
> > > > ```
> > > > 
> > > > 
> > > > I guess because I didn't do the GET? Either way this should be handled gracefully ?
> > > 
> > > 
> > > GET and PUT are on the same branch, so you can test it on this branch
> > 
> > 
> > Can you check on the PUT call, for a user who has represented marked as true, without making the GET call if he/she does the PUT call, what gets returned? For me it gives type error, if it's the same for you can you handle it gracefully maybe 403 the same way you're doing for the other case?
> 
> @meenakshi-dhanani . I don't get the same error as you did. I've got the expected 403 GET /organization not sent.
> 
> Screen Shot 2020-08-13 at 5 16 01 pm

It's because I gave OrganizationStatus as single. It breaks.

**Comment by meenakshi-dhanani** [comment]
> > > > > There's also another problem, I updated the additional info to make representative true. And it breaks with
> > > > > ```
> > > > > TypeError: Failed to fetch
> > > > > ```
> > > > > 
> > > > > 
> > > > > I guess because I didn't do the GET? Either way this should be handled gracefully ?
> > > > 
> > > > 
> > > > GET and PUT are on the same branch, so you can test it on this branch
> > > 
> > > 
> > > Can you check on the PUT call, for a user who has represented marked as true, without making the GET call if he/she does the PUT call, what gets returned? For me it gives type error, if it's the same for you can you handle it gracefully maybe 403 the same way you're doing for the other case?
> > 
> > 
> > @meenakshi-dhanani . I don't get the same error as you did. I've got the expected 403 GET /organization not sent.
> > Screen Shot 2020-08-13 at 5 16 01 pm
> 
> It's because I gave OrganizationStatus as single. It breaks.

and the timezone too. The code breaks, can we handle these as BAD REQUEST?

**Comment by mtreacy002** [comment]
> Another thing, so I'm doing the GET now and it says-Internal Server Error when I hit get,
> 
> ```
> <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
>   "http://www.w3.org/TR/html4/loose.dtd">
> <html>
>   <head>
>     <title>NameError: name 'logging' is not defined // Werkzeug Debugger</title>
>     <link rel="stylesheet" href="?__debugger__=yes&amp;cmd=resource&amp;f=style.css"
>         type="text/css">
>     
>     <link rel="shortcut icon"
>         href="?__debugger__=yes&amp;cmd=resource&amp;f=console.png">
>     <script src="?__debugger__=yes&amp;cmd=resource&amp;f=jquery.js"></script>
>     <script src="?__debugger__=yes&amp;cmd=resource&amp;f=debugger.js"></script>
>     <script type="text/javascript">
>       var TRACEBACK = 4500959440,
>           CONSOLE_MODE = false,
>           EVALEX = true,
>           EVALEX_TRUSTED = false,
>           SECRET = "hOCDEGO6cGIknxA0j847";
>     </script>
>   </head>
>   <body style="background-color: #fff">
>     <div class="debugger">
> <h1>NameError</h1>
> <div class="detail">
> ```

Mmmm.... have you tried to pull my latest code? I did make the push a couple of times. The reason I asked is because I didn't get the same error for PUT.

**Comment by mtreacy002** [comment]
> > > > > > There's also another problem, I updated the additional info to make representative true. And it breaks with
> > > > > > ```
> > > > > > TypeError: Failed to fetch
> > > > > > ```
> > > > > > 
> > > > > > 
> > > > > > I guess because I didn't do the GET? Either way this should be handled gracefully ?
> > > > > 
> > > > > 
> > > > > GET and PUT are on the same branch, so you can test it on this branch
> > > > 
> > > > 
> > > > Can you check on the PUT call, for a user who has represented marked as true, without making the GET call if he/she does the PUT call, what gets returned? For me it gives type error, if it's the same for you can you handle it gracefully maybe 403 the same way you're doing for the other case?
> > > 
> > > 
> > > @meenakshi-dhanani . I don't get the same error as you did. I've got the expected 403 GET /organization not sent.
> > > Screen Shot 2020-08-13 at 5 16 01 pm
> > 
> > 
> > It's because I gave OrganizationStatus as single. It breaks.
> 
> and the timezone too. The code breaks, can we handle these as BAD REQUEST?

Ok, will fix the error catching of Timezone and Status enum too

**Comment by mtreacy002** [comment]
Update @meenakshi-dhanani and @anitab-org/bridgeintech-maintainers . I've just pushed the latest change that addresses Meena's feedback on [error handling](https://github.com/anitab-org/bridge-in-tech-backend/pull/103#issuecomment-673313399) and [setting user cookie upon login](https://github.com/anitab-org/bridge-in-tech-backend/pull/103#issuecomment-673306038).
Note that the changes I've made here only to the main code, not test cases (which is why Travis is failing). I'm gonna work on refactoring the test cases now. It's a major change btw, so bare with me. I'll take a while.

**Comment by mtreacy002** [comment]
Update @meenakshi-dhanani and @anitab-org/bridgeintech-maintainers . I've refactored all test cases to align with the main code changes. This PR is now completed. Can you please check it out when you have time? Thanks

**Comment by meenakshi-dhanani** [comment]
It's still breaking for invalid status, can you check?

**Comment by mtreacy002** [comment]
@meenakshi-dhanani . Did you pull on my last change <commit [983c053](https://github.com/anitab-org/bridge-in-tech-backend/pull/103/commits/983c0532e73a364f4bed8fc9e854f5a405a5f648)>? Also can you show me screenshot of what's breaking? It's not clear which invalid status you're talking about. 
Invalid token, invalid input fields, or invalid data validations?

**Comment by mtreacy002** [comment]
@meenakshi-dhanani and @anitab-org/bridgeintech-maintainers . I've fixed the invalid status response and pushed the change. Can you check?
Screen Shot 2020-08-15 at 11 06 45 am

**Comment by meenakshi-dhanani** [comment]
Looks good. I checked the status, email, timezone and validations. Approving

**Review by meenakshi-dhanani — APPROVED** [review]

**Comment by meenakshi-dhanani** [comment]
@ramitsawhney27 you would need to re-approve.

**Review by ramitsawhney27 — APPROVED** [review]

**Review by ramitsawhney27 — APPROVED** [review]

**Pull request merged**
